File: frontend/app/src/main/scala/app/domain/followers/RelationShip.scala
package app.domain.followers

import app.domain.core.{AbstractEntity, Data, Link, Table, User}
import app.support.Api
import app.support.Api.given
import jfx.core.state.{ListProperty, Property}
import jfx.json.JsonMapper
import reflect.macros.ReflectMacros.reflectType

import java.util.UUID
import scala.concurrent.Future
import scala.scalajs.js
import scala.scalajs.js.JSConverters.*
import scala.scalajs.js.URIUtils.encodeURIComponent

class RelationShip extends AbstractEntity {

  val follower: Property[User | Null] = Property(null)
  val users: ListProperty[User] = ListProperty()
  val groups: ListProperty[Group] = ListProperty()

  def save(): Future[Data[RelationShip]] =
    Api.request("/service/followers/relationships/relationship").post(this).read[Data[RelationShip]]

  def update(): Future[Data[RelationShip]] =
    Api.request("/service/followers/relationships/relationship").put(this).read[Data[RelationShip]]

  def delete(): Future[Unit] =
    Api.request("/service/followers/relationships/relationship").delete(this).unit

  def updateGroups(selectedGroups: IterableOnce[Group]): Future[RelationShip] = {
    val followerId = Option(follower.get).flatMap(user => Option(user.id.get)).map(_.toString)
    followerId match {
      case Some(id) =>
        val request = new GroupAssignmentRequest(
          selectedGroups.iterator.flatMap(group => Option(group.id.get).map(_.toString)).toSeq.toJSArray
        )

        Api
          .request(s"/service/core/users/user/$id/groups")
          .put(request)
          .raw[js.Array[js.Any]]
          .map { raw =>
            groups.setAll(RelationShip.deserializeAssignedGroups(raw))
            this
          }
      case None =>
        Future.successful(this)
    }
  }
}

object RelationShip {

  def read(id: String): Future[Data[RelationShip]] =
    Api.request(s"/service/followers/relationships/relationship/$id").get.read[Data[RelationShip]]

  def list(
    index: Int,
    limit: Int,
    query: String = "",
    groups: Seq[Group] = Seq.empty,
    sorting: Seq[String] = Seq("created:desc")
  ): Future[Table[Data[RelationShip]]] = {
    val normalizedQuery = Option(query).map(_.trim).getOrElse("")
    val queryParameter =
      if (normalizedQuery.isEmpty) ""
      else s"&name=${encodeURIComponent(normalizedQuery)}"
    val sortParameter = renderSortParameters(sorting)

    val groupParameters =
      groups.iterator
        .flatMap(group => Option(group.id.get).map(_.toString))
        .map(id => s"groups=${encodeURIComponent(id)}")
        .mkString("&")

    val groupsSuffix =
      if (groupParameters.isEmpty) ""
      else s"&$groupParameters"

    Api.request(s"/service/followers/relationships?index=$index&limit=$limit$sortParameter$queryParameter$groupsSuffix").get.read[Table[Data[RelationShip]]]
  }

  private def renderSortParameters(sorting: Seq[String]): String = {
    val normalizedSorting = sorting.iterator.map(_.trim).filter(_.nonEmpty).toVector
    if (normalizedSorting.isEmpty) ""
    else normalizedSorting.map(value => s"&sort=${encodeURIComponent(value)}").mkString
  }

  private def deserializeAssignedGroups(raw: js.Array[js.Any]): Seq[Group] =
    if (raw == null || js.isUndefined(raw) || !js.Array.isArray(raw)) {
      Seq.empty
    } else {
      raw
        .iterator
        .collect {
          case value if value != null && !js.isUndefined(value) =>
            JsonMapper.deserialize(value.asInstanceOf[js.Dynamic], reflectType[Group]).asInstanceOf[Group]
        }
        .toSeq
    }
}

File: frontend/app/src/main/kotlin/app/domain/followers/RelationShip.kt
package app.domain.followers

import app.domain.core.AbstractEntity
import app.domain.core.Address
import app.domain.core.Data
import app.domain.core.Email
import app.domain.core.Link
import app.domain.core.Media
import app.domain.core.Table
import app.domain.core.User
import app.domain.core.UserInfo
import jFx2.client.JsonClient
import jFx2.state.ListProperty
import jFx2.state.ListPropertySerializer
import jFx2.state.Property
import jFx2.state.PropertySerializer
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlin.time.Clock

@Serializable
class RelationShip(
    @Serializable(with = PropertySerializer::class)
    override var id : Property<String>? = null,
    @Serializable(with = PropertySerializer::class)
    override val modified : Property<LocalDateTime> = Property(Clock.System.now().toLocalDateTime(TimeZone.UTC)),
    @Serializable(with = PropertySerializer::class)
    override val created : Property<LocalDateTime> = Property(Clock.System.now().toLocalDateTime(TimeZone.UTC)),
    @Serializable(with = PropertySerializer::class)
    val follower : Property<User> = Property(User()),
    @Serializable(with = ListPropertySerializer::class)
    val users : ListProperty<User> = ListProperty(),
    @Serializable(with = ListPropertySerializer::class)
    val groups: ListProperty<Group> = ListProperty(),
    @SerialName($$"$links")
    @Serializable(with = ListPropertySerializer::class)
    override val links : ListProperty<Link> = ListProperty()
) : AbstractEntity {

    suspend fun save() : Data<RelationShip> {
        return JsonClient.post("/service/followers/relationships/relationship", this)
    }

    suspend fun update() : Data<RelationShip> {
        return JsonClient.put("/service/followers/relationships/relationship", this)
    }

    suspend fun delete() {
        JsonClient.delete("/service/followers/relationships/relationship", this)
    }

    companion object {

        suspend fun read(id : String) : Data<RelationShip> {
            return JsonClient.invoke<Data<RelationShip>>("/service/followers/relationships/relationship/$id")
        }

        suspend fun list(index : Int, limit : Int) : Table<Data<RelationShip>> {
            return JsonClient.invoke<Table<Data<RelationShip>>>("/service/followers/relationships?index=$index&limit=$limit&sort=created:desc")
        }
    }


}